### PR TITLE
Schema as object

### PR DESCRIPTION
**Pull request by sasho-axellero**
The Schema is handled differently in the Android Wallet than in the Browser Wallet.
In the Browser Wallet it's received as Object and in the Android wallet as a String.
This update addresses that and is backwards compatible with the String Schema.

The Mobile Wallet Library (mobile_wallet_lib-release.aar) is also updated to support the new Schema as an Object format.

### CODE REVIEW

**Comment by bisgardo** [comment]
Are the Gradle changes relevant for the schema stuff? If not, please do them in a separate PR.

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerUnitTest.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import com.concordium.wallet.data.walletconnect.ParamsDeserializer
+import com.google.gson.GsonBuilder
+import org.junit.Assert
+import org.junit.Test
+
+
+class ParamsDeserializerUnitTest {
+
+    @Test
+    fun testParamsDeserializer() {
```
Comment (ali-concordium) [inline app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerUnitTest.kt:12]: The name is too generic. The name should reflect what specifically is being tested. In example - DeserializerOutputShouldBeSimilarWhenSchemaIsPassedAsObjectAndString

Also we're missing all the test scenarios for when null is returned. And a test scenario to check that deserialization actually works. If the method just returns null regardless of the input, this test will pass.

Comment (ali-concordium) [reply app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerUnitTest.kt:12]: You could also follow the Should convention, and append _Should to the test class, and have each test method follow that as a sentence that can be read. 

In example.

Class name:  ParamsDeserializer_Should
Test function: Return_Same_Output_When_Same_Schema_Is_Passed_As_String_And_Object 

Then your test name should be written as such, so you can

Comment (sasho-axellero) [reply app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerUnitTest.kt:12]: There is a class called ParamsDeserializer which methods are tested here and name reflects the naming style used in the other tests before, for example CurrencyUtilUnitTest for the CurrencyUtil., AccountDaoUnitTest for the AccountDao etc..
If null is returned, the test will fail with NullPointerException.

Comment (ali-concordium) [reply app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerUnitTest.kt:12]: Hi Sasho

The test name is too generic. It doesn't matter if it's consistently generic in other files. We should adhere to clean code principles, and have explicit names that express intent. Regardless of how it's done in different files, we need to understand what all functions (both test and non-test) do from their names. So please rename the function, and feel free to consider the options above or come up with something else, so it's very explicit what is being tested. 

I can't find any test covering that serialization is done correctly and which handles the different exception scenarios, such as JsonParseException, when deserialize function is called.

Null was a poor example, but if you modify the deserialize function to hardcore the output that is returned, no matter the input you pass to be deserialized, the unittest doesn't fail... Hence we need more tests, to test deserialization is done correctly by asserting/verifying the params.

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    fun returnNullIfJsonIsEmpty() {
+        val emptyInput = ""
+        Assert.assertNull(getParameters(emptyInput))
+    }
+
+    /**
+     * If the schema object in the request is flawed then the schema in the response should be null
+     */
+    @Test
+    fun setSchemaAsNullInResultIfSchemaObjectIsFlowed() {
```
Comment (ali-concordium) [inline app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt:84]: Flawed instead of Flowed

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import com.concordium.wallet.data.walletconnect.ParamsDeserializer
+import com.google.gson.GsonBuilder
+import com.google.gson.JsonSyntaxException
+import org.junit.Assert
+import org.junit.Test
+
+/**
+ * Unit Test for the [ParamsDeserializer]
+ */
+class ParamsDeserializerShould {
```
Comment (ali-concordium) [inline app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt:13]: Still missing a test, where the output returned from deserialize method, is verified to be correct.

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        val paramsFromObjectSchema = getParameters(paramsWithSchemaAsObject)
+
+        Assert.assertEquals(
+            paramsFromStringSchema?.schema?.value,
+            paramsFromObjectSchema?.schema?.value
+        )
+    }
+
+    /**
+     * If the parameters have invalid JSON, it should throw [JsonSyntaxException]
```
Comment (ali-concordium) [inline app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt:40]: Please remove all comments. The method names express intent making the comments redundant. Good job :)

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt (hunk 3 above)
Comment (ali-concordium) [reply app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt:13]: In example, in the arrange part of your test, define a variable named "paramWithSchemaAsString". Then make another variable named "expectedOutput". 

Then later in your test - after you deserialize "paramWithSchemaAsString", assert that paramsFromStringSchema?.schema?.value is equal to the expectedOutput. 

This should be done for both schemaAsString and schemaAsObject in two separate tests.

File: app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt (hunk 2 above)
Comment (bisgardo) [reply app/src/test/java/com/concordium/wallet/data/util/ParamsDeserializerShould.kt]: @ali-concordium There is a nice "suggest change" feature as described [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request) (item 6) that allows the author to apply the suggestion directly through the UI. And even nicer is its shortcut `ctrl-g` :)

File: app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    @Throws(JsonParseException::class)
+    override fun deserialize(
+        jElement: JsonElement,
+        typeOfT: Type?,
+        context: JsonDeserializationContext
+    ): Params {
+        val jObject = jElement.asJsonObject
+
+        return Params(
+            type = jObject["type"]?.asString,
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:19]: These `asString` calls can produce `ClassCastException` and `IllegalStateException`. These are caught below but not here. Why is that?

File: app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.concordium.wallet.data.walletconnect
+
+data class Schema (
+    val type: String?,
+    val value: String?,
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt:5]: Shouldn't these be non-nullable?

File: app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+)
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt:6]: I would suggest configuring your editor to always end files with a newline.

**Review by bisgardo — COMMENTED** [review]
Please add an explanation in the PR description of why `mobile_wallet_lib-release/mobile_wallet_lib-release.aar` is changed.

File: app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+        if (schemaElement.isJsonObject) {
+            return try {
+                val schema: Schema = context.deserialize(schemaElement, Schema::class.java)
+                if(schema.type == null || schema.value == null){
+                    return null
+                }
+                return schema
+            } catch (ex: JsonParseException) {
+                null
```
[the remark this answers is not in the document]
Comment (sasho-axellero) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:54]: The Schema is used just to print data in the UI. The old implemented method did the same, return null on error and if the result is null don't  show anything in the UI. So no error message or anything, probably thats how it was discussed and decided before, if the data is ok show it, if not don't show anything.

File: app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt [hunk 9]
```diff
@@ -297,7 +297,7 @@ class WalletConnectViewModel(application: Application) : AndroidViewModel(applic
         binder?.getSessionRequestParams()?.let { params ->
             params.payloadObj = params.parsePayload()
             params.payload = ""
-            if (params.payloadObj != null && params.payloadObj?.message != null && params.schema != null) {
+            if (params.payloadObj != null && params.payloadObj?.message != null && params.payloadObj?.receiveName != null && params.schema!= null) {
```
[the remark this answers is not in the document]
Comment (sasho-axellero) [reply app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt:300]: The 'params.payloadObj' is a mutable property that could have been changed by this time.

**Review by ali-concordium — APPROVED** [review]

File: app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt:300]: That would be a race condition that should definitely not be handled gracefully. But OK I see that this is a rule in the language. I think you should extract the value into a const. that would also alleviate the need for the `!!`s below.

Also line `params.payload = ""` above looks very suspicious. Why would we mutate the object that we should just be printing? I know this isn't your change, but it looks to me like it should be removed.

[the remark this answers is not in the document]
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/ui/walletconnect/WalletConnectViewModel.kt:300]: Btw. it's the same with the line `params.payloadObj = params.parsePayload()`. A method named `prettyPrintJson` shouldn't mutate any state.

File: app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt (hunk 6 above)
Comment (sasho-axellero) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt:5]: GSON is a java serialisation/deserialisation library that is used in this project, but there is an issue in that it doesn't recognise Kotlin's non-nullable types. So if I state them as not nullable in Kotlin, deserialising them with GSON can still produce null and issue later on.

File: app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:54]: @ali-concordium what do you think? I suggest creating a task for showing an error to the user instead of nothing.

File: app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.gson.*
+import com.google.gson.reflect.TypeToken
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonWriter
+import kotlin.jvm.internal.Reflection
+import kotlin.reflect.KClass
+import kotlin.reflect.full.memberProperties
+
+/**
+ * GSON doesn't understand Kotlin’s non-null types.
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt:12]: Please add a description of what goes wrong when this isn't added.

File: app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * This adapter adds the ability to throw an error when a non-null value is set to null.
+ * @return [TypeAdapterFactory]
+ *
+ * [JsonParseException] if a non-null value is null
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt:17]:
```suggestion
 * This adapter adds the ability to catch if a null value is assigned to a non-nullable field.
 * @return [TypeAdapterFactory]
 *
 * [JsonParseException] if null is assigned to a non-nullable field
```

File: app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+        return object : TypeAdapter<T>() {
+
+            override fun write(out: JsonWriter, value: T?) = delegate.write(out, value)
+
+            override fun read(input: JsonReader): T? {
+                val value: T? = delegate.read(input)
+
+                if (value != null) {
+                    val kotlinClass: KClass<Any> = Reflection.createKotlinClass(type.rawType)
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/core/gson/NullableTypeAdapterFactory.kt:37]: This looks potentially expensive. Can you do it just after the kotlin class check above? Then most of the work is only done once rather than on every read. You could extract the reflection objects of the nullable fields and then in `read` you only check if any of these fields are `null`.

File: app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt [hunk 13]
```diff
@@ -2,5 +2,5 @@ package com.concordium.wallet.data.walletconnect
 
 data class Schema (
     val type: String?,
-    val value: String?,
-)
\ No newline at end of file
+    val value: String?
```
Comment (bisgardo) [inline app/src/main/java/com/concordium/wallet/data/walletconnect/Schema.kt:5]: I prefer the trailing comma ;)

File: app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+        val testObject = gson.fromJson(validJsonString, TestObject::class.java)
+
+        Assert.assertEquals(
+            testObject.type,"test type"
+        )
+
+        Assert.assertEquals(
+            testObject.sender,"test sender"
+        )
```
Comment (bisgardo) [inline app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt:41]: Aren't these assertions covered by the one below?

File: app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+        Assert.assertEquals(
+            testObject, validObject
+        )
+    }
+
+    @Test
+    fun throwErrorWhenNonNullValueIsNotPresentInInput() {
```
Comment (bisgardo) [inline app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt:49]: I would find it useful to have a case where the factory isn't registered to show how it would work without it. That would also test that it actually makes a difference to register.

File: app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt (hunk 5 above)
Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:19]: I would prefer the try/catch wrapping to be in here instead to keep it more consistent. And then replace the existing try/catch with if null.

**Review by bisgardo — APPROVED** [review]

File: app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt (hunk 5 above)
Comment (sasho-axellero) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:19]: The try/catch gives us greater flexibility later on I we  want to handle the errors and show them to the user.

Comment (bisgardo) [reply app/src/main/java/com/concordium/wallet/data/walletconnect/ParamsDeserializer.kt:19]: I would agree if not for the fact that the same logic is used for the schema object, meaning that it's now in two separate places. But it's up to you.

File: app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt (hunk 15 above)
Comment (sasho-axellero) [reply app/src/test/java/com/concordium/wallet/data/util/NullableTypeAdapterFactoryShould.kt:49]: I will add another test.

**Review by bisgardo — APPROVED** [review]

**Comment by bisgardo** [comment]
> Please add an explanation in the PR description of why `mobile_wallet_lib-release/mobile_wallet_lib-release.aar` is changed.

This is not done.

**Comment by sasho-axellero** [comment]
> > Please add an explanation in the PR description of why `mobile_wallet_lib-release/mobile_wallet_lib-release.aar` is changed.
> 
> This is not done.

Sorry, I missed that. Added.
